### PR TITLE
Blog on Context Switching and its impact on performance

### PR DESCRIPTION
**Pull request by sashwathn**
Please review and approve.

### CODE REVIEW

**Review by andrewm4894 — APPROVED** [review]

File: blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+## Thrashing
+
+Thrashing is a phenomenon where the system spends more time switching between processes or threads than actually executing them. This can cause the system to become unresponsive and slow to the point of being unusable. Detecting thrashing can be challenging, but there are a few signs that can indicate its presence.
+
+One way to detect thrashing is by monitoring the system's performance metrics, such as CPU utilization, disk I/O, and memory usage. If these metrics are consistently high, while the system is slow or unresponsive, it may be an indication of thrashing.
+
+Another way to detect thrashing is by monitoring the number of context switches per second using tools like netdata. A sudden increase in context switches, especially when accompanied by a decrease in system performance, can be a sign of thrashing.
+
+In addition, monitoring the number of processes in the run queue (Runnable dimension in `system.processes_state chart`), the number of TASKLET and SCHED softirqs, as well as the time spent handling these softirqs, you can identify if there is a high level of task scheduling activity on the system.
```
Comment (cakrit) [inline blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md:28]:
```suggestion
In addition, monitoring the number of processes in the run queue via the `runnable` dimension in the `system.processes_state chart`, the number of `TASKLET` and `SCHED` softirqs, as well as the time spent handling these softirqs, you can identify if there is a high level of task scheduling activity on the system.
```

Comment (cakrit) [reply blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md:28]: I very much like how we specifically mention the chart and dimension for the first parameter. Can we do do the same for the others too?

File: blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+- **Use CPU affinity:** [CPU affinity](https://en.wikipedia.org/wiki/Processor_affinity) is a technique that assigns a specific CPU to a process or thread. This can reduce the frequency of context switching and improve cache utilization.
+
+- **Use NUMA-aware scheduling:** [NUMA (Non-Uniform Memory Access) architectures](https://en.wikipedia.org/wiki/Non-uniform_memory_access) have multiple memory nodes, and scheduling processes and threads to run on the same memory node as their data can reduce the frequency of remote memory accesses and improve performance.
+
+- **Use real-time scheduling:** Real-time scheduling policies can give priority to time-critical processes, reducing context switching and ensuring that the most important tasks are completed on time.
+
+It's worth noting that some of these options may not be appropriate for all systems and applications, and the best approach will depend on the specific requirements and constraints of your system.
```
Comment (cakrit) [inline blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md:50]: Another sentence that says nothing. Yes, obviously. If we need a disclaimer like thjis, I'd put it in the sentence before the list.

**Review by cakrit — CHANGES_REQUESTED** [review]

**Comment by sashwathn** [comment]
@cakrit : I have made the suggested changes.. I have this pending question about `runnable` state which does not exist on the `system.processes_state` chart.

**Review by cakrit — APPROVED** [review]

File: blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+Another way to detect thrashing is by monitoring the number of context switches per second using tools like netdata. A sudden increase in context switches, especially when accompanied by a decrease in system performance, can be a sign of thrashing.
+
+In addition, monitoring the number of processes in the run queue (Runnable dimension in `system.processes_state chart`), the number of TASKLET and SCHED softirqs, as well as the time spent handling these softirqs, you can identify if there is a high level of task scheduling activity on the system.
+
+In summary, detecting thrashing can be challenging, but monitoring the system's performance metrics, the number of context switches per second, the number of processes in the run queue, and the average time spent in the scheduler can help to identify its presence. By optimizing the performance of the processes or threads that are causing the most context switches, you may be able to alleviate the thrashing and improve system performance.
```
[the remark this answers is not in the document]
Comment (cakrit) [reply blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md:30]: I created a feature request for this after speaking with Costa.

**Pull request merged**